src/weights.jl

```diff
@@ -2,44 +2,202 @@
-    end
+    @compat abstract type AbstractWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T} end
 end
 
 """
-    WeightVec(vs, [wsum])
+    @weights name
 
-Construct a `WeightVec` with weight values `vs` and sum of weights `wsum`.
-If omitted, `wsum` is computed.
+Generates a new generic weight type with specified `name`, which subtypes `AbstractWeights`
+and stores the `values` (`V<:RealVector`) and `sum` (`S<:Real`).
 """
-function WeightVec{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs))
-    return WeightVec{S, eltype(vs), V}(vs, s)
+macro weights(name)
+    return quote
+        if VERSION < v"0.6.0-dev.2123"
+            immutable $name{S<:Real, T<:Real, V<:RealVector} <: AbstractWeights{S, T, V}
+                values::V
+                sum::S
+            end
+        else
+            immutable $name{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
+                values::V
+                sum::S
+            end
+        end
+    end
 end
 
+eltype(wv::AbstractWeights) = eltype(wv.values)
+length(wv::AbstractWeights) = length(wv.values)
+values(wv::AbstractWeights) = wv.values
+sum(wv::AbstractWeights) = wv.sum
+isempty(wv::AbstractWeights) = isempty(wv.values)
+
+Base.getindex(wv::AbstractWeights, i) = getindex(wv.values, i)
+Base.size(wv::AbstractWeights) = size(wv.values)
+
+"""
+    varcorrection(n::Integer, corrected=false)
+
+Compute a bias correction factor for calculating `var`, `std` and `cov` with
+`n` observations. Returns ``\\frac{1}{n - 1}`` when `corrected=true`
+(i.e. [Bessel's correction](https://en.wikipedia.org/wiki/Bessel's_correction)),
+otherwise returns ``\\frac{1}{n}`` (i.e. no correction).
+"""
+@inline varcorrection(n::Integer, corrected::Bool=false) = 1 / (n - Int(corrected))
+
+@weights Weights
+
+"""
+    Weights(vs, wsum=sum(vs))
+
+Construct a `Weights` vector with weight values `vs`.
+A precomputed sum may be provided as `wsum`.
+
+The `Weights` type describes a generic weights vector which does not support
+bias correction.
+"""
+Weights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs)) = Weights{S, eltype(vs), V}(vs, s)
+
 """
     weights(vs)
 
-Construct a `WeightVec` from a given array.
+Construct a `Weights` vector from array `vs`.
+See the documentation for [`Weights`](@ref) for more details.
 """
-weights(vs::RealVector) = WeightVec(vs)
-weights(vs::RealArray) = WeightVec(vec(vs))
+weights(vs::RealVector) = Weights(vs)
+weights(vs::RealArray) = Weights(vec(vs))
 
-eltype(wv::WeightVec) = eltype(wv.values)
-length(wv::WeightVec) = length(wv.values)
-values(wv::WeightVec) = wv.values
-sum(wv::WeightVec) = wv.sum
-isempty(wv::WeightVec) = isempty(wv.values)
+"""
+    varcorrection(w::Weights, corrected=false)
+
+Returns ``\\frac{1}{\sum w}`` when corrected is false and throws an `ArgumentError`
+when corrected is true.
+"""
+function varcorrection(w::Weights, corrected::Bool=false)
+    corrected && throw(ArgumentError("Weights does not support bias correction."))
```

Comment:
"Weights" could be ambiguous. Maybe "the Weights type does not support bias correction: use FrequencyWeights, AnalyticWeights or ProbabilityWeights if applicable".